src/dim_helpers/ConvDims.jl

```diff
@@ -6,34 +6,54 @@ export ConvDims
 Type system-level information about convolution dimensions. Critical for things like
 `im2col!()` to generate efficient code, and helpful to reduce the number of kwargs
 getting passed around.
-
-We don't want to specialize on things like image size/channel count, so we generally
-store those as fields, just for convenience, and to allow for non-breaking changes when
-we decide we _do_ want to specialize on those values.  We always want to specialize on
-things like stride, padding, dilation, and kernel flipping though.
 """
-abstract type ConvDims{N, S, P, D, F} end
+abstract type ConvDims{N} end
```

Comment:
Are we expecting the compilation improvements to mostly come from removing the type params here?

Reply:
By removing these type parameters (which were essentially runtime values) we make `Dims` as well as everything "above" that uses it type-stable.
Instead of storing these parameters in the `Dims` type we re-promote them to `Val{...}` and dispatch conv/pooling implementations on them. So in some sense, we moved these type parameters "down" the stack where they don't impact type-stability but still provide the same performance benefits.